### PR TITLE
Remove ByteBuf.readBytes(int) calls when possible

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We use ByteBuf.readBytes(int) in various places where we could either remove it completely or use readSlice(int).retain().

Modifications:

- Remove ByteBuf.readBytes(int) when possible or replace by readSlice(int).retain().

Result:

Faster code.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin PTAL

**Comment by trustin** [comment]
LGTM

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (f73ae133dcb6ba7ffa6eb75ab7f9f0073d6f0ff0) and partly into 4.0 (e35b7f982fb7cb7b181dd1ea7b1b86827c28fe41)

File: codec/src/main/java/io/netty/handler/codec/protobuf/ProtobufVarint32FrameDecoder.java [hunk 1]
```diff
@@ -60,7 +60,7 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out)
             in.resetReaderIndex();
             return;
         } else {
-            out.add(in.readBytes(length));
+            out.add(in.readSlice(length).retain());
```
Comment (johnou) [inline codec/src/main/java/io/netty/handler/codec/protobuf/ProtobufVarint32FrameDecoder.java]: readRetainedSlice right?

Comment (johnou) [inline codec/src/main/java/io/netty/handler/codec/protobuf/ProtobufVarint32FrameDecoder.java]: uh not even sure how this Netty issue came up on my radar but I just had a look at the date on it and probably predates adding this method 😆

**Pull request closed without merging**